EYBlockchain/timber
# Adding BLS12-377 MiMC Hash

## `CURVE`

Added new environment variable `CURVE` which can be set as `'BLS12_377'` or `'ALT_BN_254'` (default if blank) to be used with MiMC hashing. Not a breaking change, as curve can be left blank to use ALT-BN for MiMC.

The contracts folder was getting a bit messy, so I added `/sha` and `/mimc` folders to keep contracts in. The `MerkleTreeMiMC` and `MerkleTreeControllerMiMC` contracts in the deployer example app have been renamed with the curve type appended (e.g. `MerkleTreeMiMC_BLS12`). 

The user shouldn't have to worry about this as they are automatically deployed depending on the `CURVE` variable. Plus I added a check inside `assignDbConnection` to redirect any requests with `contractName = MerkleTreeControllerMiMC` to the relevant contract.

Tested with the `MerkleTreeController` test, roots match in soldity and node for both curve options. 

## Testing

- In `docker-compose.remote.pull.yml` set `HASH_TYPE = 'mimc'` and `CURVE = 'BLS12_377'` for both merkle-tree and deployer services
- Run `docker-compose -f docker-compose.remote.pull.yml up --build`
- Start the event filter by sending a post request to `http://localhost:9000/start` with body `{
	"contractName": "MerkleTreeControllerMiMC"
}` (`MerkleTreeControllerMiMC_BLS12` would also work)
- Run `docker-compose -f docker-compose.remote.pull.yml run --rm deployer npx mocha --exit --require @babel/register 'test/MerkleTreeController.test.js'` to run the test
- Update the tree and get metadata by sending a post request to `http://localhost:9000/update` with body `{
	"contractName": "MerkleTreeControllerMiMC"
}`
- Confirm that the root and frontier match the test logs!
